### ISSUE TITLE
Persist font color from the outside

### ISSUE THREAD

**Issue by jdujava** [T-F-S/tcolorbox#305]
I stumbled upon a situation where I wanted to use current font color also inside `tcolorbox` box. It seems that the color is always reset (by default to black), and I couldn't find anything in documentation about how to "persist" the outside font color.

I was able to come up with a workaround (below is included a working example for anyone interested).
While it is not complicated, perhaps having something similar supported directly by `tcolorbox` could be even handier.

```latex
\documentclass{article}

\usepackage{codehigh} % provides \fakeverb (which can be used inside other commands)
\usepackage{tcolorbox}

\tcbset{
    macro/.style={
        nobeforeafter, box align=base, boxsep=0em, before upper={\vphantom{]}},
        left=0.15em, right=0.15em, top=0.25ex, bottom=0.20ex,
        colframe=black!15, colback=black!3, fontupper=\ttfamily,
        arc=0.1em, rounded corners=all, boxrule=0.3mm,
        grow sidewards by=0.05em, enlarge bottom by=-0.05ex, enlarge top by=-0.4ex,
    },
}
\NewTCBox{\macrobox}{!O{}}{macro,#1}
\NewDocumentCommand{\macro}{s O{} m}{% starred version uses current text color also inside the box
    \IfBooleanTF{#1}{%
        \colorlet{currentcolor}{.}% save current color
        \macrobox[coltext=currentcolor,#2]{\fakeverb{#3}}%
    }{%
        \macrobox[#2]{\fakeverb{#3}}%
    }%
}

\begin{document}
To write \LaTeX{} use \tcboxverb{\LaTeX} \fakeverb{\LaTeX} \macro{\LaTeX} \macro*{\LaTeX}.

\color{gray} 
To write \LaTeX{} use \tcboxverb{\LaTeX} \fakeverb{\LaTeX} \macro{\LaTeX} \macro*{\LaTeX}.

\vspace{2ex} \color{black} 
Now using \macro{\textcolor} --- \textcolor{red}{red \macro{\LaTeX} \macro*{\LaTeX}}, \textcolor{blue}{blue \macro{\LaTeX} \macro*{\LaTeX}}.
\end{document}
```
![image](https://github.com/user-attachments/assets/5d3ab118-486a-46b6-9615-bb89d8b9451e)

**Comment by muzimuzhi** [comment]
Just set `coltext=.` directly?

```tex
\documentclass{article}

\usepackage{codehigh} % provides \fakeverb (which can be used inside other commands)
\usepackage{tcolorbox}

\tcbset{
    macro/.style={
        nobeforeafter, box align=base, boxsep=0em, before upper={\vphantom{]}},
        left=0.15em, right=0.15em, top=0.25ex, bottom=0.20ex,
        colframe=black!15, colback=black!3, fontupper=\ttfamily,
        arc=0.1em, rounded corners=all, boxrule=0.3mm,
        grow sidewards by=0.05em, enlarge bottom by=-0.05ex, enlarge top by=-0.4ex,
    },
}
\NewTCBox{\macrobox}{!O{}}{macro,#1}

\begin{document}
normal \macrobox[coltext=.]{\fakeverb{\LaTeX}} normal

\textcolor{red}{red \macrobox[coltext=.]{\fakeverb{\LaTeX}} red}

\textcolor{blue}{blue \macrobox[coltext=.]{\fakeverb{\LaTeX}} blue}
\end{document}
```

![image](https://github.com/user-attachments/assets/56e8a4ae-f062-41a5-9304-d41470805a6d)

**Comment by jdujava** [comment]
Oh wow, that works?! I assumed that this wouldn't see the outside color, so I didn't even try.
Thanks @muzimuzhi for suggestion!

If one wants this behavior by default, `coltext=.` can be even added to `macro/.style={...}` (I tested and it works nicely).

**Comment by jdujava** [comment]
Note, that to set it globally, one should use
```latex
\tcbset{every box/.style = {coltext=.}}
```
Just using
```latex
\tcbset{coltext=.}
```
didn't work when I tried.

Do I understand correctly that the second one was evaluated immediately (and only once) to the current color (just black), while the first is evaluated at the time a box is created (and thus leads to the desired behavior)?

**Comment by jdujava** [comment]
Thinking more about this: Shouldn't — when no explicit `coltext` (or something equivalent) is set in a given style, use the current font color — be the default?

Or is there some reason to switch to black by default?

**Comment by muzimuzhi** [comment]
> Do I understand correctly that the second one was evaluated immediately (and only once) to the current color (just black), while the first is evaluated at the time a box is created (and thus leads to the desired behavior)?

Yes. Unfortunately this subtle difference can only be told from their implementations.

`coltext=<color>` is equivalent to `colupper=<color>,collower=<color>` (this step _is_ documented), and the current implementation of `colupper` and `collower` will call an immediate  (this detail is _not_ documented) `\colorlet`.

https://github.com/T-F-S/tcolorbox/blob/b59111a7429eccf196e716edec170f52fe560232/tex/latex/tcolorbox/tcolorbox.sty#L298-L300
https://github.com/T-F-S/tcolorbox/blob/b59111a7429eccf196e716edec170f52fe560232/tex/latex/tcolorbox/tcolorbox.sty#L153

`every box/.style={<kv-list>}` only stores `<kv-list>` in `every box` style key (without even parsing the `<kv-list>`, a `pgfkeys` implementation detail). The style key is then used for every `tcolorbox` boxes, and at that time the `<kv-list>` is parsed and "evaluated".

> Shouldn't — when no explicit `coltext` (or something equivalent) is set in a given style, use the current font color — be the default?
> 
> Or is there some reason to switch to black by default?

Well it's more of a (historic) design decision which can only be answered by the author of `tcolorbox` @T-F-S.  Meanwhile for backwards compatibility, maybe it's too late to change their initial values.

It seems feasible to add a new option (or a set of) for skipping the color setting for title, upper and lower parts.

**Comment by T-F-S** [comment]
> Note, that to set it globally, one should use
> 
> ```
> \tcbset{every box/.style = {coltext=.}}
> ```
> 
> Just using
> 
> ```
> \tcbset{coltext=.}
> ```
> 
> didn't work when I tried.
> 
> Do I understand correctly that the second one was evaluated immediately (and only once) to the current color (just black), while the first is evaluated at the time a box is created (and thus leads to the desired behavior)?

Yes, this is correct:
```latex
\tcbset{coltext=.}
```
immediately applies the content, while
```latex
\tcbset{every box/.style = {coltext=.}}
```
defines a style in the `tikz` manor which is applied when the box(es) is(are) used.

**Comment by T-F-S** [comment]
> Well it's more of a (historic) design decision which can only be answered by the author of `tcolorbox` @T-F-S. Meanwhile for backwards compatibility, maybe it's too late to change their initial values.
> 
> It seems feasible to add a new option (or a set of) for skipping the color setting for title, upper and lower parts.

Well, the package is called `tcolorbox` with the purpose to created **colored** boxes. It was planned as somewhat fancier but similar to `\colorbox` or `\fcolorbox`. Somehow, the original purpose became a little bit out of hand over the years ;-)

I think the solutions found above are a good way to achieve the goal to have a persistent font color. I am not planning to add customization to skip color settings for texts or backgrounds or frames.
